predicate/predicate_paging.go

```diff
@@ -0,0 +1,260 @@
+func PagingWithPredicate(internalPredicate Predicate, pageSize int32) *predPaging {
+	return &predPaging{
+		internalPredicate: internalPredicate,
+		comparator:        nil,
+		iterationType:     ENTRY,
+		pageSize:          pageSize,
+		page:              0,
+		anchorList:        []types.Entry{},
+	}
+}
+
+// IterationType allows the paging predicate to be used to fetch values, a keyset or an entryset.
+type IterationType string
+
+const (
+	KEY   IterationType = "KEY"
+	VALUE               = "VALUE"
+	ENTRY               = "ENTRY"
+)
+
+func iterationTypeToId(t IterationType) int8 {
+	switch t {
+	case KEY:
+		return 0
+	case VALUE:
+		return 1
+	case ENTRY:
+		return 2
+	default:
+		return -1
+	}
+}
+
+// NextPage sets the page of the paging predicate to the next page to fetch the next results in the following operation.
+func (p *predPaging) NextPage() {
+	p.page++
+}
+
+// PrevPage sets the page of the paging predicate to the previous page to fetch the previous results in the following operation.
+func (p *predPaging) PrevPage() {
+	p.page--
+}
+
+// SetPage sets the page to a given page number to fetch the results at that page in the following operation.
+func (p *predPaging) SetPage(page int32) {
+	p.page = page
+}
+
+// GetPage return the current page of the paging predicate.
+func (p predPaging) GetPage() int32 {
+	return p.page
+}
+
+// AnchorList returns the anchor list of the paging predicate containing the first item of every page fetched with the predicate.
+func (p predPaging) AnchorList() []types.Entry {
+	return p.anchorList
+}
+
+func (p predPaging) FactoryID() int32 {
+	return factoryID
+}
+
+func (p predPaging) ClassID() int32 {
+	return 15
+}
+
+func (p predPaging) ReadData(input serialization.DataInput) {
+	p.internalPredicate = input.ReadObject().(Predicate)
+	p.comparator = input.ReadObject().(serialization.IdentifiedDataSerializable)
+	p.page = input.ReadInt32()
+	p.pageSize = input.ReadInt32()
+	p.iterationType = IterationType(input.ReadByte())
+	size := input.ReadInt32()
+	p.anchorList = make([]types.Entry, size)
+	for i := int32(0); i < size; i++ {
+		page := input.ReadInt32()
+		key := input.ReadObject()
+		value := input.ReadObject()
+		anchorEntry := types.Entry{Key: key, Value: value}
+		p.anchorList[i] = types.Entry{Key: page, Value: anchorEntry}
+	}
+}
+
+func (p predPaging) WriteData(output serialization.DataOutput) {
+	output.WriteObject(p.internalPredicate)
+	output.WriteObject(p.comparator)
+	output.WriteInt32(p.page)
+	output.WriteInt32(p.pageSize)
+	output.WriteString(string(p.iterationType))
+	output.WriteInt32(int32(len(p.anchorList)))
+	for _, a := range p.anchorList {
+		output.WriteInt32(a.Key.(int32))
+		anchorEntry := a.Value.(types.Entry)
+		output.WriteObject(anchorEntry.Key)
+		output.WriteObject(anchorEntry.Value)
+	}
+}
+
+func (p predPaging) String() string {
+	return fmt.Sprintf("paging, predicate=%s, pageSize=%d, page=%d",
```

Comment:
How about returning something like (`Paging(predicate=%s, ...)` to be consistent with other predicates?

Reply:
Addressed in d226845